server/channels/api4/user.go

```diff
@@ -3443,3 +3445,74 @@ func getUsersWithInvalidEmails(c *Context, w http.ResponseWriter, r *http.Reques
 		c.Logger.Warn("Error writing response", mlog.Err(err))
 	}
 }
+
+func getUsersForReporting(c *Context, w http.ResponseWriter, r *http.Request) {
+	if !(c.IsSystemAdmin() && c.App.SessionHasPermissionTo(*c.AppContext.Session(), model.PermissionSysconsoleReadUserManagementUsers)) {
+		c.SetPermissionError(model.PermissionSysconsoleReadUserManagementUsers)
+		return
+	}
+
+	sortColumn := "Username"
+	if r.URL.Query().Get("sort_column") != "" {
+		sortColumn = r.URL.Query().Get("sort_column")
+	}
+
+	sortDesc := false
+	if r.URL.Query().Get("sort_direction") == "desc" {
+		sortDesc = true
+	}
+
+	pageSize := 50
+	if pageSizeStr, err := strconv.ParseInt(r.URL.Query().Get("page_size"), 10, 64); err == nil {
+		pageSize = int(pageSizeStr)
+	}
+
+	lastSortColumnValue := r.URL.Query().Get("last_column_value")
+	lastUserId := r.URL.Query().Get("last_id")
+	roleFilter := r.URL.Query().Get("role_filter")
+	hasNoTeam := r.URL.Query().Get("has_no_team") == "true"
+	teamFilter := r.URL.Query().Get("team_filter")
+	if !(teamFilter == "" || model.IsValidId(teamFilter)) {
+		c.Err = model.NewAppError("getUsersForReporting", "api.getUsersForReporting.invalid_team_filter", nil, "", http.StatusBadRequest)
+		return
+	}
+	hideActive := r.URL.Query().Get("hide_active") == "true"
+	hideInactive := r.URL.Query().Get("hide_inctive") == "true"
+
+	startAt := int64(0)
+	endAt := int64(0)
+	currentTime := time.Now()
+	dateRangeStr := r.URL.Query().Get("date_range")
+	if dateRangeStr == "last30days" {
+		startAt = currentTime.AddDate(0, 0, -30).UnixMilli()
+	} else if dateRangeStr == "previousmonth" {
+		startOfMonth := time.Date(currentTime.Year(), currentTime.Month(), 1, 0, 0, 0, 0, time.Local)
+		startAt = startOfMonth.AddDate(0, -1, 0).UnixMilli()
+		endAt = startOfMonth.UnixMilli()
+	} else if dateRangeStr == "last6months" {
+		startAt = currentTime.AddDate(0, -6, -0).UnixMilli()
+	}
+
+	userReports, err := c.App.GetUsersForReporting(
+		sortColumn,
+		sortDesc,
+		pageSize,
+		lastSortColumnValue,
+		lastUserId,
+		startAt,
+		endAt,
+		roleFilter,
+		teamFilter,
+		hasNoTeam,
+		hideActive,
+		hideInactive,
```

Comment:
This is starting to become unreadable now. The recommended approach is to pack them into a single struct and pass that instead. You can see how the search APIs are written along with the various search filters.

Reply:
+1. This will cleanup a lot of places.

Reply:
Is there a reason we do this? I figured method signatures and IntelliSense make this work basically the same. I can see a bit more in this case because we are able to omit some properties as optional but I don't think it really saves us much and just means that every parameter goes from being `parameter,` to `Parameter: parameter,`

Reply:
Thanks Devin. @streamer45 had a discussion on this exact thing a long time back in a developer's meeting. Claudio, could you perhaps distil some of that discussion here? Happy to move this to a dedicated thread in community if this is going out of scope of the PR.